rfcs/0042-bitswap.md

```diff
@@ -0,0 +1,268 @@
+<!-- 
+- TODO:
+  - call out inefficiency in design caused by only having a single writer to the Bitswap block cache
+    - WAIT: I think this needs addressed
+    - options:
+      - 2 writer db model
+      - writer proxy process
+-->
+
+## Summary
+[summary]: #summary
+
+This RFC proposes adding Bitswap to our libp2p networking stack in order to address issues related to our current gossip network pub/sub layer.
+
+## Motivation
+[motivation]: #motivation
+
+Mina has very large messages that are broadcast over the gossip net pub/sub layer. This incurs a high bandwidth cost due to the nature of our pub/sub rebroadcast cycles work in order to consistently broadcast messages throughout the network. For example, we observer blocks on mainnet as large as 15mb. This would represent only a single block, and each block broadcast message has a multiplicative cost on bandwidth as it's being broadcast throughout the network. This bandwidth cost also translates into CPU cost due to the cost of hashing incoming messages to check against the de-duplication cache before processing them. We currently observe behaviors where the libp2p helper process can be pegged at 100% CPU on certain hardware setups when there is high pub/sub throughput on the network. Since gossip pub/sub is used not only for blocks, but also transactions and snark work, the broadcasting of each of these simultaneously ends up compounding the issue.
+
+Implementing Bitswap in our libp2p layer will address this issue by allowing us to immediately reduce our pub/sub message size, while making the larger data referenced by pub/sub messages available upon request. It provides a mechanism for breaking up large data into chunks that can be distributed throughout the network (streamed back from multiple peers), and a system for finding peers on the network who are able to serve that data.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+[Bitswap](https://docs.ipfs.io/concepts/bitswap/) is a module provided by [libp2p](https://libp2p.io/) that enables distributed data synchronization over a p2p network, somewhat comparable to how [BitTorrent](https://en.wikipedia.org/wiki/BitTorrent) works. It works by splitting up data into chunks called blocks (we will explicitly refer to these as "Bitswap blocks" to disambiguate them from "blockchain blocks"), which are structured into a tree or DAG with a single root. When a node on the network wants to download to some data, it asks it's peers to see which (if any) have the root Bitswap block corresponding to that data. If none of the peers have the data, it falls back to querying the gossip network's [DHT](https://docs.ipfs.io/concepts/dht/#kademlia) to find a suitable node that can serve the data.
```

Comment:
According to the [go implementation of Bitswap](https://github.com/ipfs/go-bitswap/blob/master/docs/how-bitswap-works.md), Bitswap first does the following things in parallel:

1. Sends a `want-have` to all peers in the current "Bitswap session" (a subset of connected peers Bitswap prioritizes based on recent Bitswap related activity)  to see if they have the block.
2. Chooses a random peer in the current "Bitswap session" and sends a `want-block` instead of a `want-have`. If that peer does not have the block, it moves onto another random peer in the session.

Bitswap does this until either it gets the block back from a `want-block` message, learns of a peer that has a block from a `want-have` message, or it hears back from all peers that they `DONT_HAVE` the block. If the latter occurs, meaning no peers in the session have the block that's needed, it then does the following in parallel:

1. Sends a `want-have` to all other connected peers.
2. Queries the DHT to find a peer that definitely has the block.

At this point in time, once it finds a peer in the DHT that has the block, if it still has not received a `HAVE` message from any of the peers it is connected to, then it will establish a new connection to the peer it looked up in the DHT (if it wasn't a peer we were already connected to), and sends a `want-block` to it then. Keep in mind that, if any of the peers whom initially told us they `DONT_HAVE` block receives the block while this is happening, they will send us a `HAVE` message letting us know they now can serve that block to us. So, in effect, the node will only establish a connection to a new node it finds via the DHT if no connected peers receive the block by the time this described algorithm has executed, but it's hard to quantify the exact delay.

Should I include this writeup in the RFC?

Reply:
> Presumably there'll be a gossip of a new block and all other nodes will attempt to hit that peer as soon as they discover it

With the above description, when a new block header is broadcast on the network, it's first published to the connected peers of the block producer. Those peers will use that existing connection to download the block body via Bitswap, adding the block producer's node to their "Bitswap session", and initially only downloading the related Bitswap blocks from that node since it is the only node that has the data yet. Simultaneously, these nodes that just received the block header via gossip will forward the block header onto their connected peers. These 2nd peers will ask their peers who has the related blocks, and will only establish a new connection back to the block producer (assuming they didn't have one already) if none of their peers receive the related blocks before they finish executing the algorithm described above and have determined no peers have the related blocks yet.

Reply:
> Should I include this writeup in the RFC?

Yes please.

We chatted a bit about this out-of-band, where I realised that gossiping the root bitswap block ID for a mina block should be done separately, and this gossip is only propagated once the peers have downloaded and verified the data in those bitswap blocks. I think it would be good to call this detail out: originally, I was confused into thinking that we would broadcast the new block's bitswap block ID to the whole network at once, and was concerned about the whole network simultaneously trying to connect to the source node, it being the only source of that bitswap block.

For valid blocks, it would be nicer (ie. faster to propagate blocks round the network) if we could send the gossip indiscriminately. However, I don't know if bitswap has a mitigation for this 'stampede to the source' issue, and this also means that every peer in the network could end up downloading the header of a bad block before banning the peer that originated it, which seems like a potential DoS vector.